### PR TITLE
binlog decode implement

### PR DESCRIPTION
**Pull request by chaychen2005**

### CODE REVIEW

File: libstorage/BinLogHandler.cpp [hunk 1]
```diff
@@ -219,4 +219,200 @@ void BinLogHandler::encodeBlock(
...
+        entry->setID(id);
+        uint8_t status = *((uint8_t*)&buffer[offset]);
+        offset++;
+        entry->setStatus(status);
+        BINLOG_HANDLER_LOG(TRACE) << LOG_DESC("entry info") << LOG_KV("idx", idx)
+                                  << LOG_KV("id", entry->getID())
+                                  << LOG_KV("status", (uint32_t)entry->getStatus());
+        for (const auto& key : vecField)
+        {
+            if (key == "_status_" || key == "_num_" || key == "_id_")
```
Comment (bxq2011hust) [inline libstorage/BinLogHandler.cpp]:
```suggestion
            if (key == STATUS || key == NUM_FIELD || key == ID_FIELD)
```

File: libstorage/BinLogHandler.cpp [hunk 2]
```diff
@@ -219,4 +219,200 @@ void BinLogHandler::encodeBlock(
...
+bool BinLogHandler::getBlockData(
+    BinLogInfo& binlog, int64_t dataNum, std::map<int64_t, std::vector<TableData::Ptr>>& blocksData)
+{
+    while (binlog.offset < binlog.length)
+    {
+        bytes buffer;
+        // get block data length
+        buffer.resize(sizeof(uint32_t));
+        uint32_t size =
+            binlog.file.read(reinterpret_cast<char*>(buffer.data()), buffer.size()).gcount();
```
Comment (bxq2011hust) [inline libstorage/BinLogHandler.cpp]:
```suggestion
uint32_t length=0;
            binlog.file.read(reinterpret_cast<char*>(&length), sizeof(uint32_t)).gcount();
```

File: libstorage/BinLogHandler.h [hunk 3]
```diff
@@ -60,7 +81,7 @@ class BinLogHandler
     /// @return true : it's completely written
     /// @return false : missing binlog in database
     bool getMissingBlocksFromBinLog(
-        int64_t num, std::map<int64_t, std::vector<TableData::Ptr>> binLogData);
+        int64_t dataNum, std::map<int64_t, std::vector<TableData::Ptr>>& binLogData);
```
Comment (bxq2011hust) [inline libstorage/BinLogHandler.h]:
```suggestion
        int64_t _dataNum, std::map<int64_t, std::vector<TableData::Ptr>>& _binLogData);
```

File: libstorage/BinLogHandler.h [hunk 4]
```diff
@@ -40,6 +40,27 @@ namespace storage
...
+{
+    Success = 0,               // decode block success
+    WrittenBlock = 1,          // the block has already been wirtten
+    BlockDataLengthError = 2,  // decode block fail because of invalid block data length
+};
+
+class BinLogInfo
+{
+public:
+    BinLogInfo(const std::string& filePath) : version(0), length(0), offset(0)
```
Comment (bxq2011hust) [inline libstorage/BinLogHandler.h]: why need this struct?

File: libstorage/BinLogHandler.cpp (hunk 1 above)
Comment (chaychen2005) [reply libstorage/BinLogHandler.cpp]: Done.

File: libstorage/BinLogHandler.h (hunk 3 above)
Comment (chaychen2005) [reply libstorage/BinLogHandler.h]: Done.

File: libstorage/BinLogHandler.cpp (hunk 2 above)
Comment (chaychen2005) [reply libstorage/BinLogHandler.cpp]: Done.

File: libstorage/BinLogHandler.h [hunk 5]
```diff
@@ -40,6 +40,27 @@ namespace storage
...
 #endif
 
+enum DecodeBlockResult
+{
+    Success = 0,               // decode block success
+    WrittenBlock = 1,          // the block has already been wirtten
+    BlockDataLengthError = 2,  // decode block fail because of invalid block data length
+};
+
+class BinLogInfo
```
[the remark this answers is not in the document]
Comment (chaychen2005) [reply libstorage/BinLogHandler.h]: Done.

File: libstorage/BinLogHandler.h (hunk 4 above)
Comment (chaychen2005) [reply libstorage/BinLogHandler.h]: manage Binlog context used in functions

File: libstorage/BinLogHandler.h [hunk 6]
```diff
@@ -40,6 +40,26 @@ namespace storage
...
+    Success = 0,               // decode block success
+    WrittenBlock = 1,          // the block has already been wirtten
+    BlockDataLengthError = 2,  // decode block fail because of invalid block data length
+};
+
+struct BinLogContext
+{
+    BinLogContext(const std::string& path) : version(0), length(0), offset(0)
+    {
+        file.open(path, std::ios::in | std::ios::binary);
```
Comment (morebtcg) [inline libstorage/BinLogHandler.h]: consider open failed

Comment (chaychen2005) [reply libstorage/BinLogHandler.h]: Judge when using binlog context. If fail, exit function.

**Pull request merged**